core/trino-main/src/main/java/io/trino/operator/aggregation/arrayagg/GroupArrayAggregationState.java

```diff
@@ -13,35 +13,110 @@
+        extends AbstractGroupedAccumulatorState
         implements ArrayAggregationState
 {
-    private static final int MAX_BLOCK_SIZE = 1024 * 1024;
-    private static final int VALUE_CHANNEL = 0;
+    private static final int INSTANCE_SIZE = instanceSize(GroupArrayAggregationState.class);
+
+    // See java.util.ArrayList for an explanation
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    private final FlatArrayBuilder arrayBuilder;
+    private long[] groupHeadPositions = new long[0];
+    private long[] groupTailPositions = new long[0];
+
+    public GroupArrayAggregationState(Type type, MethodHandle readFlat, MethodHandle writeFlat)
+    {
+        arrayBuilder = new FlatArrayBuilder(type, readFlat, writeFlat, true);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return INSTANCE_SIZE +
+                sizeOf(groupHeadPositions) +
+                sizeOf(groupTailPositions) +
+                arrayBuilder.getEstimatedSize();
+    }
+
+    @Override
+    public void ensureCapacity(long maxGroupId)
+    {
+        checkArgument(maxGroupId + 1 < MAX_ARRAY_SIZE, "Maximum array size exceeded");
+        int requiredSize = toIntExact(maxGroupId + 1);
+        if (requiredSize > groupHeadPositions.length) {
+            int newSize = Ints.constrainToRange(requiredSize * 2, 1024, MAX_ARRAY_SIZE);
+            int oldSize = groupHeadPositions.length;
+
+            groupHeadPositions = Arrays.copyOf(groupHeadPositions, newSize);
+            Arrays.fill(groupHeadPositions, oldSize, newSize, -1);
+
+            groupTailPositions = Arrays.copyOf(groupTailPositions, newSize);
+            Arrays.fill(groupTailPositions, oldSize, newSize, -1);
+        }
+    }
 
-    GroupArrayAggregationState(Type valueType)
+    @Override
+    public void addAll(Block block)
     {
-        super(PageBuilder.withMaxPageSize(MAX_BLOCK_SIZE, ImmutableList.of(valueType)));
+        for (int position = 0; position < block.getPositionCount(); position++) {
+            add(block, position);
+        }
     }
 
     @Override
     public void add(Block block, int position)
     {
-        prepareAdd();
-        appendAtChannel(VALUE_CHANNEL, block, position);
+        int groupId = (int) getGroupId();
+        long index = arrayBuilder.size();
+
+        if (groupTailPositions[groupId] == -1) {
+            groupHeadPositions[groupId] = index;
+        }
+        else {
+            arrayBuilder.setNextIndex(groupTailPositions[groupId], index);
+        }
+        groupTailPositions[groupId] = index;
+        arrayBuilder.add(block, position);
+    }
+
+    @Override
+    public void writeAll(BlockBuilder blockBuilder)
+    {
+        long nextIndex = getGroupHeadPosition();
+        checkArgument(nextIndex != -1, "Group is empty");
+        while (nextIndex != -1) {
+            nextIndex = arrayBuilder.write(nextIndex, blockBuilder);
+        }
```

Comment:
It might be more efficient to add a `writeAll` to `FlatArrayBuilder` that handles the linked structure. The `write` method does a lot of work that could be hoisted out of this loop (checkIndex, getRecords, getRecordOffset) and the JIT may not be smart enough to do it on its own.